File: MauiMilan/Core/Domain/Battle/BattleSimulator.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Milan.Domain.Battle
{
    public class BattleSimulator
    {
        readonly Random _rng;
        public BattleSimulator(Random rng) { _rng = rng; }

        public BattleResult Simulate(UnitStats[] teamA, UnitStats[] teamB, int maxTurns)
        {
            teamA ??= System.Array.Empty<UnitStats>();
            teamB ??= System.Array.Empty<UnitStats>();
            var a = teamA.Select(u => new S { Stats = u, Hp = u.Hp, A = true }).ToList();
            var b = teamB.Select(u => new S { Stats = u, Hp = u.Hp, A = false }).ToList();

            for (int turn = 1; turn <= maxTurns; turn++)
            {
                var order = a.Where(x => x.Hp > 0)
                    .Concat(b.Where(x => x.Hp > 0))
                    .OrderByDescending(x => x.Stats.Spd)
                    .ThenBy(x => _rng.Next())
                    .ToList();

                foreach (var actor in order)
                {
                    if (actor.Hp <= 0) continue;
                    var enemies = actor.A ? b : a;
                    var target = enemies.Where(e => e.Hp > 0).OrderBy(e => e.Hp).FirstOrDefault();
                    if (target == null || target.Hp <= 0) continue;
                    // 伤害公式唯一事实来源：与 BattleActivity 手动出牌走同一入口。
                    target.Hp -= StrikeDamage(actor.Stats, target.Stats);
                }

                // 空队伍无法"全部死亡"，必须要求队伍非空，否则空 teamB 会因 LINQ 语义被误判为胜利。
                if (b.Count > 0 && b.All(x => x.Hp <= 0))
                    return Done(true, turn, a, b);
                if (a.Count > 0 && a.All(x => x.Hp <= 0))
                    return Done(false, turn, a, b);
            }

            return Done(false, maxTurns, a, b);
        }

        static BattleResult Done(bool victory, int turns, List<S> a, List<S> b) => new BattleResult
        {
            Victory = victory,
            Turns = turns,
            RemainingHp = a.Sum(x => Math.Max(0, x.Hp)),
            OpponentRemainingHp = b.Sum(x => Math.Max(0, x.Hp)),
        };

        /// <summary>单体攻击结算伤害（Simulate 与手动出牌共用，单一事实来源）。
        /// 攻方属性由 GameState.ComputeStats 生成，已含等级/突破/天赋/升星的加成。</summary>
        public static int StrikeDamage(UnitStats attacker, UnitStats defender)
            => System.Math.Max(1, attacker.Atk - defender.Def / 2);

        class S { public UnitStats Stats; public int Hp; public bool A; }
    }
}

File: MilanKotlin/app/src/main/java/com/milan/game/domain/battle/BattleSimulator.kt
package com.milan.game.domain.battle

import kotlin.random.Random

/**
 * 回合制战斗模拟器（C# Milan.Domain.Battle.BattleSimulator 翻译）。
 */
class BattleSimulator(private val rng: Random) {

    /**
     * 模拟 [maxTurns] 回合的战斗：
     * - 行动顺序：存活单位按 Spd 降序，同速随机（rng）决定先后；
     * - 目标选择：存活敌人中剩余血量最少者（先集火残血）；
     * - 空队伍不参与胜负判定（防空敌队被误判为胜利）；
     * - 回合数耗尽未分胜负按失败返回。
     */
    fun simulate(teamA: Array<UnitStats>?, teamB: Array<UnitStats>?, maxTurns: Int): BattleResult {
        val a = (teamA ?: emptyArray()).map { S(it, it.hp, true) }
        val b = (teamB ?: emptyArray()).map { S(it, it.hp, false) }

        for (turn in 1..maxTurns) {
            val order = a.filter { it.hp > 0 }
                .plus(b.filter { it.hp > 0 })
                .sortedWith(compareByDescending<S> { it.stats.spd }.thenBy { rng.nextInt() })

            for (actor in order) {
                if (actor.hp <= 0) continue
                val enemies = if (actor.a) b else a
                val target = enemies.filter { it.hp > 0 }.minByOrNull { it.hp }
                if (target == null || target.hp <= 0) continue
                // 伤害公式唯一事实来源：与手动出牌走同一入口。
                target.hp -= strikeDamage(actor.stats, target.stats)
            }

            // 空队伍无法"全部死亡"，必须要求队伍非空，否则空 teamB 会被误判为胜利。
            if (b.isNotEmpty() && b.all { it.hp <= 0 }) return done(true, turn, a, b)
            if (a.isNotEmpty() && a.all { it.hp <= 0 }) return done(false, turn, a, b)
        }

        return done(false, maxTurns, a, b)
    }

    private fun done(victory: Boolean, turns: Int, a: List<S>, b: List<S>): BattleResult = BattleResult(
        victory = victory,
        turns = turns,
        remainingHp = a.sumOf { it.hp.coerceAtLeast(0) },
        opponentRemainingHp = b.sumOf { it.hp.coerceAtLeast(0) },
    )

    companion object {
        /** 单体攻击结算伤害（simulate 与手动出牌共用，单一事实来源）。
         * 攻方属性由 GameState.ComputeStats 生成，已含等级/突破/天赋/升星的加成。 */
        fun strikeDamage(attacker: UnitStats, defender: UnitStats): Int =
            (attacker.atk - defender.def / 2).coerceAtLeast(1)
    }

    private class S(val stats: UnitStats, var hp: Int, val a: Boolean)
}
